### PR TITLE
PER-8345: Manage Tags/Archive Settings

### PR DESCRIPTION
**Pull request by meisekimiu**
## Description
This PR adds the manage tags functionality to the web-app, located in a new dialog called "Archive Settings", currently available to Archive Owners (and eventually Managers but the back end doesn't support that right now). This dialog lets you edit and delete existing tags in the current archive, allowing you to rename a tag on all records in the archive or delete an entire tag as well.

To hopefully help get this approved ASAP I'm going to preemptively explain a few of the specific changes made in this PR:
- Jest configuration is changed since the Manage-Tags tests are the first tests in the `CoreModule`, a specifically problematic module in our codebase for Jest Testing. The config changes help us ignore some errors from using some specific Typescript libraries that were not included in the previous Jest-compatible tests written in the `OnboardingModule`.
- The change to `src/app/notifications/services/notification.service.ts` is written for the same reason. Jest does not like the NodeJS.Timeout type. (But seriously, isn't this a browser application? Why are we using a builtin NodeJS type there anyway? Is that a thing you can just... do?)
- The introduction of the `TagVO` class in #63 requires the renaming of some methods that use the similarly shaped interface `TagVOData` but call them "TagVOs". You can see this renaming scheme in `src/app/shared/services/api/tag.repo.ts` but any files that call those methods are modified as well.
- The changes in `src/app/file-browser/components/edit-tags/edit-tags.component.ts` are to refresh the currently viewed item's tags when you update a tag in the new dialog. To test this out, select a record in My Files so you can see its metadata and tags list, then edit one of the tags in the Manage Tags dialog. With the changes listed, you should see the file's tag list instantly update instead of needing to be refreshed manually.
- I checked in a tiny amount of commented out code in `src/app/core/components/archive-settings-dialog/archive-settings-dialog.component.ts` because currently the backend does not allow managers to edit/delete tags, but that will be fixed eventually, and we can uncomment the code with that fix.

## Steps to Test
This functionality has been tested by QA already, but here's a thorough checklist:
- Open the Archive Settings and edit/delete tags
- Verify a confirmation box is shown when trying to delete a tag
- Verify the tag filter search thing works correctly
- Begin renaming a tag and then cancel out of editing it. The tag name *should* revert. Make sure to test this in conjunction with saving other tags.
- Click a record in My Files so you can see its metadata and tags list, and then open archive settings and update one of that record's tags. When the dialog closes, you should see the item's tag is updated immediately without the need to manually refresh.
- Test that you cannot use the Manage-Tags component without the proper access role.
- Test the Archive Settings dialog in mobile view.
- Make sure I didn't leave any `console.log`s lying around.

### CODE REVIEW

**Review by cecilia-donnelly — COMMENTED** [review]
Leaving some initial questions here now for you; about to actually test how it works!

@meisekimiu this is clearly so much work!!!  Thank you thank you for thinking all this through, writing such clear code, and adding test coverage.  Way to go!

File: src/app/core/components/manage-tags/manage-tags.component.spec.ts
Comment (jasonaowen) [inline src/app/core/components/manage-tags/manage-tags.component.spec.ts:121]: Having these `expect`s in the middle of the test method feels a bit strange to me. The structure I learned for tests was [arrange, act, assert](http://wiki.c2.com/?ArrangeActAssert), and the advantage of following that pattern is that tests become easier to read and easier to think about. Extra asserts in the middle of a method tend to mean that more tests would be better; here, I think we would benefit from, perhaps, "should have an edit button" and "should enter edit mode".

File: src/app/core/components/manage-tags/manage-tags.component.spec.ts
Comment (jasonaowen) [inline src/app/core/components/manage-tags/manage-tags.component.spec.ts:137]: And given that we have a few copies of this now, and perhaps more to come, it might be worth extracting a function like `enterEditMode`.

File: src/app/core/components/manage-tags/manage-tags.component.spec.ts
Comment (jasonaowen) [inline src/app/core/components/manage-tags/manage-tags.component.spec.ts:148]: Let's also break this up into multiple tests. It looks like each instance of this is the same, with only the input and expected count differing; I believe in Jest that can most easily be done with the [`each` method](https://jestjs.io/docs/api#testeachtablename-fn-timeout).

File: src/app/core/components/manage-tags/manage-tags.component.ts
Comment (jasonaowen) [reply src/app/core/components/manage-tags/manage-tags.component.ts:33]: That summary sounds right to me! And the code appears to be pretty idiomatic. What would you suggest for a docstring?

File: src/app/core/components/manage-tags/manage-tags.component.ts
Comment (jasonaowen) [reply src/app/core/components/manage-tags/manage-tags.component.ts:47]: The `ManageTagsComponent` keeps a list of its tags, and this line removes the deleted tag from that list, by filtering the list to all tags that do not match the deleted tag's ID.

File: src/app/core/components/manage-tags/manage-tags.component.ts
Comment (jasonaowen) [inline src/app/core/components/manage-tags/manage-tags.component.ts:51]: If tag deletion fails, we should put the deleted tag back into the list. And probably log an error to Sentry. And maybe display an error message.

File: src/app/models/tag-vo.ts
Comment (jasonaowen) [reply src/app/models/tag-vo.ts:18]: Allowing global tags would involve some front-end work, including updating this. I think unless and until that happens, the API should not return tags with a null `archiveId`, and the front-end should expect it not to be null.

(I also think that we should enforce that at the database level until that global tags feature is worked on, but that's a bit out-of-scope for this pull request.)

File: src/app/core/components/manage-tags/manage-tags.component.ts
Comment (jasonaowen) [inline src/app/core/components/manage-tags/manage-tags.component.ts:17]: I think we want a [map](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map) instead of an array here. For example, editing a tag with a `tagId` of 1234 would, I think, resize this array to size 1234 (or really, 1235, because zero-indexed)... and the later `delete` will not affect the size.

Example from repl:
```
$ node
Welcome to Node.js v14.18.3.
Type ".help" for more information.
> x = [1, 2, 3]
[ 1, 2, 3 ]
> x[5] = 5
5
> x[5]
5
> x
[ 1, 2, 3, <2 empty items>, 5 ]
> x.length
6
> delete x[5]
true
> x
[ 1, 2, 3, <3 empty items> ]
> x.length
6
```

File: src/app/core/components/archive-settings-dialog/archive-settings-dialog.component.ts
Comment (jasonaowen) [inline src/app/core/components/archive-settings-dialog/archive-settings-dialog.component.ts:40]: Same comment about Sentry and retry limit / exponential backoff.

File: src/app/file-browser/components/edit-tags/edit-tags.component.ts
Comment (jasonaowen) [inline src/app/file-browser/components/edit-tags/edit-tags.component.ts:167]: Could this be rewritten more simply with [`Array.find`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/find)?

```suggestion
    this.itemTags = this.item.TagVOs.map(
      (tag) => this.allTags.find((t) => t.tagId === tag.tagId)
    );
```

Alternatively, it might make more sense to store these in `Map`s rather than `Array`s, if this is a frequent operation.

**Review by jasonaowen — DISMISSED** [review]
This is great, @meisekimiu! As @cecilia-donnelly, it's clear you put a ton of work into this, and I really appreciate the coherent commits. Also echoing the +1 for test coverage - nicely done!

**Comment by cecilia-donnelly** [comment]
Just one comment from testing in the app:

Love that the archive list on an item refreshes immediately when I edit.  Can you make that happen on delete, too?  That is, I deleted an archive in the modal but still see it on an item until I refresh the page.

Other than that everything worked the way I expected!  Thanks so much!

File: src/app/core/components/manage-tags/manage-tags.component.ts
Comment (cecilia-donnelly) [reply src/app/core/components/manage-tags/manage-tags.component.ts:33]: It took me a bit to grok what was happening because this uses "filter" instead of "search" in the method name (which I believe is because it's using a pattern called "filter" and is correct!) so I would appreciate a hint that points the casual reader in the right direction.  "Search tags for user-provided string" or something.  BUT that's just me, no need to block on it.

File: src/app/core/components/manage-tags/manage-tags.component.ts
Comment (cecilia-donnelly) [reply src/app/core/components/manage-tags/manage-tags.component.ts:47]: Ah, so this means the frontend doesn't need to do a road trip to exclude that deleted tag.  Nice!  Thanks for the explanation, @jasonaowen .

File: src/app/models/tag-vo.ts
Comment (cecilia-donnelly) [reply src/app/models/tag-vo.ts:18]: Fair enough.

**Comment by jasonaowen** [comment]
Let's not block on this, but I suspect the data wrangling this code needs to do would be a useful lens for thinking about the permanent-sdk: what library API would be most helpful for getting this specific work done? Some of the complexity here is specific to the front end, but some of it will be shared by other clients, and this might be a great time to think about that.

File: src/app/core/components/archive-settings-dialog/archive-settings-dialog.component.ts
Comment (jasonaowen) [reply src/app/core/components/archive-settings-dialog/archive-settings-dialog.component.ts]: Wait did I make the same comment on the same chunk of code twice? How did I manage to do that? Sorry!

**Comment by jasonaowen** [comment]
We mentioned during standup wanting to get a partial review done on the progress so far. These three commits (7efac42 0f2ecc8 0ac95d3) look good to me! I went through and marked the corresponding conversations as resolved. Thanks, @meisekimiu!

We also discussed reviewers, and depending on timing and availability, it might be @cecilia-donnelly that finishes the review. If so, of course feel free to dismiss my review!

File: src/app/core/components/archive-settings-dialog/archive-settings-dialog.component.ts
Comment (meisekimiu) [reply src/app/core/components/archive-settings-dialog/archive-settings-dialog.component.ts:62]: Yes! Since there's only one tab in the dialog right now, we don't have to do anything when the user clicks on a tab since there are no other ones to change to. I already implemented the function because I mainly copy-pasted the HTML structure from our other dialogs into this new archive settings dialog, including calls to `setTab`.

File: src/app/core/components/archive-settings-dialog/archive-settings-dialog.component.ts
Comment (meisekimiu) [reply src/app/core/components/archive-settings-dialog/archive-settings-dialog.component.ts:65]: This was added because when initially testing on dev, tags were not getting their ArchiveId set by the backend. This was extra insurance in case the back end doesn't return the Ids, although that initial problem was caused by the feature flag not being turned on. 

(That does means it can probably be removed with no issues)

File: jest.config.js
Comment (meisekimiu) [reply jest.config.js]: This was something I realized, but I initially was dealing with a rather absurd problem: somehow adding that string caused some Jest errors to disappear even though we didn't use any of these packages. I tried modifying the line and it would cause the errors to appear again so I kept the line like that and continued working on it.

Upon removing that line now it still works, like one would expect it to in the first place. Which is good but I'm not entirely sure why that line ever fixed anything earlier in the process. Anyway, line is removed now!

File: src/app/core/components/left-menu/left-menu.component.html
Comment (cecilia-donnelly) [reply src/app/core/components/left-menu/left-menu.component.html:92]: Gotcha, yes, good to be consistent.  Thanks!

**Review by cecilia-donnelly — APPROVED** [review]
Thank you for your careful work, @meisekimiu !

**Pull request merged**
